src/app.js

```diff
@@ -1,80 +1,160 @@
-        return cat.id == catInQuestion;         // interchangeable with cat['id']
-    }); 
-    if (catIndex !== -1) {
-        req.catIndex = catIndex;
-        next()
+const CatRepository = require('./catsDb.js'); // class instance / object  
+const breeds = require('./breedsDb.js');
+const { deleteCatByIndex } = require('./catsDb.js');
+
+
+// simple JS function 
+const isInvalidString = (value) => {
+    if (typeof value === "string") {
+        return false;
+    } else if (typeof value === "undefined") {
+        return false;                     
+    };                                   
+}
+// if value is undefined, then we can let it be ""; 
+
+// simple JS function 
+const generateErr = (value) => {
+    let message =  `Property value "${value}" is not a valid string`;
+    const newError = new Error(message);
+    newError.status = 404;
+    return newError; 
+}
+
+// checks id type 
+const isIdNum = (req, res, next) => {
+    console.log('id being verified');
+    const id = Number(req.params.id);                   // extracts id as a number - returns NaN if it can't 
+    if (id) {   
+        req.id = id;                                       // Alt: if (typeOf(id) === "number") { ... }        
+        console.log('id verified');
+        next() 
     } else {
-        res.status(404).send('cat does not exist in this database');
+        const newError =  new Error(`cat id '${id}' is not a valid number`)
+        newError.status = 404;
+        next(newError); 
     };
 }
 
-const checkObject = (req, res, next) => {
-    if (req.query) {
-        next()
-    } else {
-        res.status(400).send('invalid information provided')
-    }
+// checks object keys are valid 
+const checkObjKeys = (req, res, next) => { 
+    const possibleKeys = ["name", "sex", "coat", "description", "breedId"]; 
+    const objToCheck = req.query;                // { name: "", sex: "", coat: ""}
+    arrayOfKeys = Object.keys(objToCheck);      // Eg: [ "name", "sex", "coat"]
+    arrayOfKeys.forEach((key) => { 
+        console.log(`checking key: ${key}`)
+        const index = possibleKeys.indexOf(key);    // alternative is to use .findByIndex(callback)
+        if (index < 0) {
+            let message =  `"${key}" is not a valid property`;
+            const newError = new Error(message);
+            newError.status = 404;
+            return next(newError); 
+        } else {
+        console.log(`key '${key}' verified`)
+        }
+    });
+    console.log('object successfully checked');
+    req.object = objToCheck; 
+    next();
 }
 
-let nextId = 5;
+// checks the object values are valid 
+const checkObjValues = (req, res, next) => { 
+    const objToCheck = req.query;      // { name: "", sex: "", coat: ""}
 
-// set up a port
-// app.use('/path', callback) to mount middleware functions (default called on request received) 
+    if (isInvalidString(objToCheck.name)) {     
```

Comment:
Looking good :) you might be able to simplify this even further for some of the fields which use the same validation:

```js
const invalidStringKeys = ["name", "sex", "coat", ...];

invalidStringKeys.forEach { key ->
    const obj = objToCheck[key];

    if (isInvalidString(obj)) {
        const err = generateErr(obj);
        return next(err);
    }
}
```